### ISSUE TITLE
Support for LLVM 18 fuzzing

### ISSUE THREAD

**Issue by Darksonn** [google/oss-fuzz#11626]
The Tokio build on oss-fuzz has started failing with the following output:
```
Running fuzz_linked_list
Running fuzz_stream_map
Traceback (most recent call last):
  File "/usr/local/bin/profraw_update.py", line 129, in <module>
    sys.exit(main())
  File "/usr/local/bin/profraw_update.py", line 120, in main
    profraw_latest = upgrade(profraw_base, sect_prf_cnts, sect_prf_data)
  File "/usr/local/bin/profraw_update.py", line 87, in upgrade
    relativize_address(data, offset + 16, dataref, sect_prf_cnts, sect_prf_data)
  File "/usr/local/bin/profraw_update.py", line 35, in relativize_address
    value = struct.unpack('Q', data[offset:offset + 8])[0]
struct.error: unpack requires a buffer of 8 bytes
warning: /workspace/out/libfuzzer-coverage-x86_64/dumps/fuzz_stream_map.16753151517152022731_0.profraw: unsupported instrumentation profile format version
error: no profile can be merged
[2024-02-21 06:24:21,463 INFO] Finding shared libraries for targets (if any).
[2024-02-21 06:24:21,473 INFO] Finished finding shared libraries for targets.
error: fuzz_stream_map: Failed to load coverage: No such file or directory
error: Could not load coverage information
error: No such file or directory: Could not read profile data!
Traceback (most recent call last):
  File "/usr/local/bin/profraw_update.py", line 129, in <module>
    sys.exit(main())
  File "/usr/local/bin/profraw_update.py", line 120, in main
    profraw_latest = upgrade(profraw_base, sect_prf_cnts, sect_prf_data)
  File "/usr/local/bin/profraw_update.py", line 87, in upgrade
    relativize_address(data, offset + 16, dataref, sect_prf_cnts, sect_prf_data)
  File "/usr/local/bin/profraw_update.py", line 35, in relativize_address
    value = struct.unpack('Q', data[offset:offset + 8])[0]
struct.error: unpack requires a buffer of 8 bytes
warning: /workspace/out/libfuzzer-coverage-x86_64/dumps/fuzz_linked_list.14910788477734546202_0.profraw: unsupported instrumentation profile format version
error: no profile can be merged
```
This error seems very similar to #6268, and the latest Rust release has the following in its changelog:

> [Update the minimum external LLVM to 16.](https://github.com/rust-lang/rust/pull/117947/)

Let me know if the problem is on our end, or if there's anything I can do to help.

For more information, see [the chromium bug for the build failure](https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=66904).

**Comment by seanmonstar** [comment]
Also with httparse.

**Comment by jonathanmetzman** [comment]
OK, I will do a roll this week or next depending on when my latest clusterfuzz work ends

**Comment by Xuanwo** [comment]
There is another failure from https://github.com/apache/opendal/issues/4242

**Comment by jonathanmetzman** [comment]
Sorry, I'm putting out some fires and will need to get back to this, later this week.

**Comment by oliverchang** [comment]
ryu also seems to  be failing for a similar reason: https://oss-fuzz-build-logs.storage.googleapis.com/log-acc3b39e-2aa0-4769-a499-4c224201e35e.txt

**Comment by DavidKorczynski** [comment]
Proposed a fix here https://github.com/google/oss-fuzz/pull/11681 downgrading Rust which can be used until LLVM 18 upgrade lands.

**Comment by jonathanmetzman** [comment]
We've upgraded to a recent clang. @DavidKorczynski Should we undo your change.

**Comment by DavidKorczynski** [comment]
I'm not sure, from the perspective one of the things I did was pin Rust to a specific version: https://github.com/google/oss-fuzz/blob/6b05655f197d1e36849cfb501730e2391ae358cf/infra/base-images/base-builder/install_rust.sh#L18

Rust was the only language that wasn't pinned to a specific build, so I thought maybe this is something we'd like to keep moving forward?

**Comment by kaczmarczyck** [comment]
[OpenSK](https://github.com/google/OpenSK) has the same problem, but recently, also our fuzzing workflow broke.
I tried different Rust compiler versions between November 2023 and today, and all fail the same way.
Is that related to this issue, or should I open a new one?

Example failing run:
https://github.com/google/OpenSK/actions/runs/9254285343/job/25455744646

**Comment by kaczmarczyck** [comment]
Confirming this fixed the build problem:
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=69138#c4

**Comment by alexcrichton** [comment]
I've attempted to review the current status of this and other various related updates and I've submitted https://github.com/google/oss-fuzz/pull/12075 to update the pinned Rust nightly to today's nightly. That being said if there are still remaining blockers I'm not aware of I can try to help hunt those down or otherwise fix our build issues via other means.

**Comment by alexcrichton** [comment]
I saw yeah but I haven't tried tro reproduce and I wasn't sure how applicable that was. Given that both seem to be using LLVM 18 I was assuming it might be stale by this point. Is there a way to see the failure that happens locally?

As for updating projects. I can grep for "2023-12-28" which has a number of hits and I can try to update those projects, but when you mentionto bump clang/llvm where is that? I thought it was 

https://github.com/google/oss-fuzz/blob/7f915006751d37347078beb0040fb41c0d2a3436/infra/base-images/base-clang/checkout_build_install_llvm.sh#L52-L53

which already looked like llvm 18

**Comment by catenacyber** [comment]
Late to the party, but what is the status here ? cc @maflcko cf https://github.com/google/oss-fuzz/pull/12077#issuecomment-2236780291
(coming from https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=70505 )

In case it helps : I took a look into https://reviews.llvm.org/D138846 and made https://github.com/catenacyber/oss-fuzz/tree/profraw9 :
It translates/downgrades profraw version 9 to version 8 so that current clang `llvm-profdata merge` can read nightly-rust profraw files.
But then, llvm-cov fails because there was also a bump for the elf section with profile stuff (INSTR_PROF_COVMAP_VERSION). 
I can do it the other way (and update profraw8 to 9)

**Comment by catenacyber** [comment]
So, the alternative to llvm update to match rust, would be to patch profraw_update.py but also create a new program that updates the compiled fuzz targets covmap section to remove this new bitmap information...

TL;DR Updating llvm seems the right way to go

By the way, I am not sure profraw_update.py is needed anymore for profraw version 5 (swift ? )

**Comment by catenacyber** [comment]
> Agree. However, some legacy projects will have their coverage build broken by that.

So updating profraw_update.py may help there

**Comment by catenacyber** [comment]
See https://github.com/google/oss-fuzz/pull/12365
